### PR TITLE
Allow plugins to override default options

### PR DESCRIPTION
**Pull request by czosel**
This is a first attempt to handle #3972. See https://github.com/prettier/plugin-php/pull/52 for a usage example. I attached `defaultOptions` to the printer to reuse existing logic (`getPrinter`). Like that, only files handled by the respective plugin should be affected by the changed defaults.

This doesn't handle multiparser support, though - but maybe that's something to be handled in the future when there is a concrete use case?

### CODE REVIEW

**Comment by azz** [comment]
I think we want the API to be

```js
exports.defaultOptions = {...}
```

Rather than:

```js
exports.printers.php.defaultOptions = {...}
```

Because we may want to show this information in the CLI. e.g.:

```shellsession
$ prettier --help tab-width
--tab-width <int>

  Number of spaces per indentation level.

Default: 2
Plugin defaults:
* @prettier/plugin-php: 4
```

Do you have any thoughts on this @ikatyang?

**Comment by czosel** [comment]
That makes sense! I just pushed some changes that add the plugin's options to a new property `pluginOptions` in `supportOptions`, which should make it easy to display them in the CLI's help. I read from the same object in `options.js` to extract the current defaults. 

In this implementation a plugin's default options are registered under the name of it's first print - which works because currently all plugins only have one printer. Maybe there's a better solution, for example a `name` property on the plugin?

If you think this is generally the right way to move forward, i'll continue by cleaning up the (currently a bit clumsy) implementation and add some tests!

File: src/common/support.js [hunk 1]
```diff
@@ -249,6 +249,17 @@ function getSupportInfo(version, opts) {
...
 
       return newOption;
+    })
+    .map(option => {
+      const filteredPlugins = plugins.filter(
+        plugin => plugin.defaultOptions && plugin.defaultOptions[option.name]
+      );
+      const pluginDefaults = filteredPlugins.reduce((reduced, plugin) => {
+        const printerName = Object.keys(plugin.printers)[0];
+        reduced[printerName] = plugin.defaultOptions[option.name];
```
Comment (ikatyang) [inline src/common/support.js]: We probably need to attach the plugin name (package name, plugin filepath, or the `name` property if provided) in `load-plugins.js` and use it here as an ID.

https://github.com/prettier/prettier/blob/9643d3370233f2870c43b81d0855833fb983d9f4/src/common/load-plugins.js#L6

File: src/main/options.js [hunk 2]
```diff
@@ -55,6 +55,21 @@ function normalize(options, opts) {
   rawOptions.locStart = parser.locStart;
   rawOptions.printer = getPrinter(rawOptions);
 
+  const pluginDefaults = supportOptions
+    .filter(
+      optionInfo =>
+        optionInfo.pluginDefaults && optionInfo.pluginDefaults[options.parser]
```
Comment (ikatyang) [inline src/main/options.js]: Same here, since the parser name might not equal to the printer name.

**Comment by czosel** [comment]
@ikatyang Right - i updated the implementation.

File: src/cli/util.js [hunk 3]
```diff
@@ -613,7 +613,13 @@ function createDetailedUsage(context, optionName) {
       ? `\n\nDefault: ${createDefaultValueDisplay(optionDefaultValue)}`
       : "";
 
-  return `${header}${description}${choices}${defaults}`;
+  const pluginDefaults =
+    option.pluginDefaults && Object.keys(option.pluginDefaults).length
+      ? `\nPlugin defaults:${Object.keys(option.pluginDefaults).map(
+          key => `\n* ${key}: ${option.pluginDefaults[key]}`
```
Comment (ikatyang) [inline src/cli/util.js]: We should call `createDefaultValueDisplay(value)` here, though there is no non-primitive value in core options.

File: src/main/options.js [hunk 4]
```diff
@@ -53,7 +53,23 @@ function normalize(options, opts) {
...
+    )
+    .reduce(
+      (reduced, optionInfo) =>
+        Object.assign(reduced, {
+          [optionInfo.name]: optionInfo.pluginDefaults[plugin.name]
+        }),
+      {}
+    );
+
+  Object.assign({}, defaults, pluginDefaults);
```
[the remark this answers is not in the document]
Comment (czosel) [reply src/main/options.js]: Oops, thats embarassing :sweat_smile: Probably still have to figure out how to rewrite it that i don't have to use `let` (and test is also still missing)

File: tests_integration/__tests__/early-exit.js [hunk 5]
```diff
@@ -26,6 +26,26 @@ describe(`show detailed usage with --help l (alias)`, () => {
...
 
+describe(`show detailed usage with plugin options`, () => {
+  runPrettier("cli", [
+    "--help",
+    "tab-width",
+    "--plugin=../plugins/automatic/node_modules/prettier-plugin-bar",
+    "--parser=bar"
+  ]).test({
+    status: 0,
+    stdout: `--tab-width <int>
```
[the remark this answers is not in the document]
Comment (czosel) [reply tests_integration/__tests__/early-exit.js]: Of course - didn't think that doing that would be so easy :+1:

File: src/main/options.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ikatyang) [reply src/main/options.js]: You can just assign it to a new variable, e.g. `mixedDefaults` or something else:

```js
const mixedDefaults = Object.assign({}, defaults, pluginDefaults);

// Object.keys(mixedDefaults) ...
```

File: tests_integration/__tests__/__snapshots__/early-exit.js.snap [hunk 6]
```diff
@@ -514,6 +514,21 @@ exports[`show detailed usage with --help write (stdout) 1`] = `
...
+exports[`show detailed usage with plugin options (stderr) 1`] = `""`;
+
+exports[`show detailed usage with plugin options (stdout) 1`] = `
+"--tab-width <int>
+
+  Number of spaces per indentation level.
+
+Default: 2
+Plugin defaults:
+* ../plugins/automatic/node_modules/prettier-plugin-bar: 4
```
Comment (azz) [inline tests_integration/__tests__/__snapshots__/early-exit.js.snap]: Could we use the `package.json` name here? (Happy to do this in a follow up PR though)

Comment (ikatyang) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: That's the result from `--plugin=../plugins/automatic/node_modules/prettier-plugin-bar`, I guess it should be fine?

- `--plugin=path/to/plugin.js` (path)
   
  ```
  Plugin defaults:
  * path/to/plugin.js: 4
  ```

- `--plugin=prettier-plugin-bar` (package name)
   
  ```
  Plugin defaults:
  * prettier-plugin-bar: 4
  ```

Comment (azz) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: Ah, cool. I'm happy with that. Does it work for plugins that are automatically resolved?

Comment (ikatyang) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: It should work. @czosel could you add a test for that? just clone the test case and set its cwd to `path/to/plugins/automatic` then remove the `--plugin=xxx` part.

Comment (czosel) [reply tests_integration/__tests__/__snapshots__/early-exit.js.snap]: done!

File: src/main/options.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (czosel) [reply src/main/options.js]: right - done.

File: tests_integration/plugins/automatic/node_modules/prettier-plugin-bar/index.js [hunk 7]
```diff
@@ -19,5 +19,8 @@ module.exports = {
     bar: {
       print: path => concat(["bar+", path.getValue().text])
     }
+  },
+  defaultOptions: {
```
Comment (azz) [inline tests_integration/plugins/automatic/node_modules/prettier-plugin-bar/index.js]: @ikatyang mentioned this above, but could you add an integration test to ensure that this `tabWidth` is applied?

**Comment by azz** [comment]
LGTM other than missing an integration test.

File: src/main/options.js [hunk 8]
```diff
@@ -53,14 +53,30 @@ function normalize(options, opts) {
...
+      {}
+    );
 
   Object.keys(defaults).forEach(k => {
     if (rawOptions[k] == null) {
       rawOptions[k] = defaults[k];
     }
   });
 
+  Object.assign(rawOptions, pluginDefaults);
```
Comment (czosel) [inline src/main/options.js]: Just realized that `rawOptions` already contains the default options, so assigning the "mixedDefaults" only if the option key is not present in rawOptions (line 73) won't work. I added some tests for it.

File: tests_integration/plugins/defaultOptions/plugin.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+"use strict";
+
+module.exports = {
+  languages: [
+    {
+      name: "foo",
+      parsers: ["foo-parser"],
+      extensions: [".foo"],
+      since: "1.0.0"
```
Comment (azz) [inline tests_integration/plugins/defaultOptions/plugin.js]: Can delete `since` now.

File: src/main/options.js (hunk 8 above)
Comment (czosel) [reply src/main/options.js]: Hmm, if I'm not mistaken the order of option application should be
```
core defaults < plugin defaults < cli / file overrides
```
That would mean that we already have to find out which plugin is used before CLI options are applied, so the logic has to be moved to somewhere around here:

 https://github.com/prettier/prettier/blob/master/src/cli/util.js#L224

File: src/cli/util.js [hunk 10]
```diff
@@ -835,11 +844,13 @@ function updateContextOptions(context, plugins) {
...
-        Object.assign(reduced, { [optionInfo.name]: optionInfo.default }),
-      Object.assign({}, optionsModule.hiddenDefaults)
-    );
+    .reduce((reduced, optionInfo) => {
+      // default are applied in main/options.js when pluginDefaults are present
+      if (Object.keys(optionInfo.pluginDefaults).length) {
+        return reduced;
+      }
+      return Object.assign(reduced, { [optionInfo.name]: optionInfo.default });
+    }, Object.assign({}, optionsModule.hiddenDefaults));
```
Comment (czosel) [inline src/cli/util.js]: I managed to get the tests passing by adding this workaround - but I'm not really happy with it. I could imagine that there's a case where this breaks, and even if there isn't it doesn't look very clean to me.

I'd be glad if someone could explain the bigger picture on the way defaults are handled right now: To me, the two-step process of applying defaults seems confusing:

1. CLI defaults are applied in `cli/util.js`
https://github.com/prettier/prettier/blob/9ee113cd2e66dad5567ce03170d1c6c6dff1a310/src/cli/util.js#L226
2. Defaults are applied if an option is `null` in `main/options.js` 
https://github.com/prettier/prettier/blob/9ee113cd2e66dad5567ce03170d1c6c6dff1a310/src/main/options.js#L58

I saw that there are some use cases where the core options are still `null` in step two, but I'm not sure what those use cases are, and consequently how the proper process of applying defaults should look like in the future.

Comment (ikatyang) [reply src/cli/util.js]: --> #4045

**Comment by czosel** [comment]
Workaround is reverted, tests are passing :tada:

File: docs/plugins.md [hunk 11]
```diff
@@ -149,6 +155,33 @@ function embed(
...
+```js
+options: {
+  openingBraceNewLine: {
+    type: "boolean",
+    category: "Global",
+    default: true,
+    description: "Move open brace for code blocks onto new line."
+  }
+}
+```
```
Comment (j-f1) [inline docs/plugins.md]: Can you document each key in the option?

File: docs/plugins.md [hunk 12]
```diff
@@ -149,6 +155,33 @@ function embed(
...
+
+### `defaultOptions`
+
+If your plugin requires different default values for some of prettier's core options, you can specify them in `defaultOptions`:
+
+```
```
Comment (j-f1) [inline docs/plugins.md]: ` ```js `?

File: docs/plugins.md (hunk 11 above)
Comment (czosel) [reply docs/plugins.md]: Of course! Can you help my out with the `category` option? I'll do this in a new PR.

**Pull request merged**